src/lib/xmlparse.rs

```diff
@@ -1968,76 +1975,81 @@ unsafe extern "C" fn destroyBindings(mut bindings: *mut BINDING, mut parser: XML
+                let mut p: *mut TAG = 0 as *mut TAG;
+                if tagList.is_null() {
+                    if self.m_freeTagList.is_null() {
+                        break;
+                    }
+                    tagList = self.m_freeTagList;
+                    self.m_freeTagList = NULL as *mut TAG
+                }
+                p = tagList;
+                tagList = (*tagList).parent;
+                FREE!(self, (*p).buf as *mut c_void);
+                destroyBindings((*p).bindings, self);
             }
-            tagList = (*parser).m_freeTagList;
-            (*parser).m_freeTagList = NULL as *mut TAG
-        }
-        p = tagList;
-        tagList = (*tagList).parent;
-        FREE!(parser, (*p).buf as *mut c_void);
-        destroyBindings((*p).bindings, &mut *parser);
-    }
-    /* free m_openInternalEntities and m_freeInternalEntities */
-    entityList = (*parser).m_openInternalEntities;
-    loop {
-        let mut openEntity: *mut OPEN_INTERNAL_ENTITY = 0 as *mut OPEN_INTERNAL_ENTITY;
-        if entityList.is_null() {
-            if (*parser).m_freeInternalEntities.is_null() {
-                break;
+            /* free m_openInternalEntities and m_freeInternalEntities */
+            entityList = self.m_openInternalEntities;
+            loop {
+                let mut openEntity: *mut OPEN_INTERNAL_ENTITY = 0 as *mut OPEN_INTERNAL_ENTITY;
+                if entityList.is_null() {
+                    if self.m_freeInternalEntities.is_null() {
+                        break;
+                    }
+                    entityList = self.m_freeInternalEntities;
+                    self.m_freeInternalEntities = NULL as *mut OPEN_INTERNAL_ENTITY
+                }
+                openEntity = entityList;
+                entityList = (*entityList).next;
+                FREE!(self, openEntity as *mut c_void);
+            }
+            destroyBindings(self.m_freeBindingList, self);
+            destroyBindings(self.m_inheritedBindings, self);
+            poolDestroy(&mut self.m_tempPool);
+            poolDestroy(&mut self.m_temp2Pool);
+            FREE!(self, self.m_protocolEncodingName as *mut c_void);
+            /* external parameter entity parsers share the DTD structure
+            parser->m_dtd with the root parser, so we must not destroy it
+            */
+            if self.m_isParamEntity == 0 && !self.m_dtd.is_null() {
+                /* XML_DTD */
+                dtdDestroy(
+                    self.m_dtd,
+                    self.m_parentParser.is_null() as XML_Bool,
+                    &self.m_mem,
+                );
+            }
+            FREE!(self, self.m_atts as *mut c_void);
+            FREE!(self, self.m_groupConnector as *mut c_void);
+            FREE!(self, self.m_buffer as *mut c_void);
+            FREE!(self, self.m_dataBuf as *mut c_void);
+            FREE!(self, self.m_nsAtts as *mut c_void);
+            let _ = self.m_unknownEncoding.take();
```

Comment:
I wonder if these are needed. After the outer drop is called, Rust recursively calls the field drops. Since you're not freeing the structure at the end anymore, it should be safe to just leave this to the field drops.